lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/RestRequestExecutor.cs

```diff
@@ -20,73 +20,95 @@
-            IRestRequest request,
+            RestRequest request,
             Uri requestUri,
-            CancellationToken cancellationToken) where T : new()
+            CancellationToken cancellationToken)
         {
-            var client = _clientFactory.CreateRestClient(requestUri);
-
-            var response =
-                await
-                    client.ExecuteTaskAsync<T>(request, cancellationToken);
-
-            if (response.ErrorException != null)
-            {
-                throw new YarnRestAPIException("Executing REST API failed", response.ErrorException);
-            }
-
+            RestResponse<T> response;
             try
             {
-                // HTTP status code greater than 300 is unexpected here.
-                // See if the server sent a error response and throw suitable
-                // exception to user.
-                if (response.StatusCode >= HttpStatusCode.Ambiguous)
-                {
-                    Log.Log(Level.Error, "RESTRequest failed. StatusCode: {0}; Response: {1}", response.StatusCode, response.Content);
-                    var errorResponse = JsonConvert.DeserializeObject<Error>(response.Content);
-                    throw new YarnRestAPIException { Error = errorResponse };
-                }
+                response = await _client.ExecuteRequestAsync<T>(request, requestUri, cancellationToken);
             }
             catch (Exception exception)
             {
-                throw new YarnRestAPIException("Unhandled exception in deserializing error response.", exception);
+                throw new YarnRestAPIException("Unhandled exception in executing REST request.", exception);
             }
 
+            HandleResponse(response.Exception, response.StatusCode, response.Content);
+
             return response.Data;
         }
 
-        public async Task<IRestResponse> ExecuteAsync(IRestRequest request, Uri uri, CancellationToken cancellationToken)
+        public async Task<RestResponse<VoidResult>> ExecuteAsync(RestRequest request,
+            Uri requestUri,
+            CancellationToken cancellationToken)
         {
-            var client = _clientFactory.CreateRestClient(uri);
-
+            RestResponse<VoidResult> response;
             try
             {
-                return await client.ExecuteTaskAsync(request, cancellationToken);
+                response = await _client.ExecuteRequestAsync(request, requestUri, cancellationToken);
             }
             catch (Exception exception)
             {
                 throw new YarnRestAPIException("Unhandled exception in executing REST request.", exception);
             }
+
+            HandleResponse(response.Exception, response.StatusCode, response.Content);
+            return response;
+        }
+
+        private static void HandleResponse(Exception responseException, HttpStatusCode httpStatusCode, string content)
+        {
+            if (responseException != null)
```

Comment:
Is there any way for the user to get the status code for the Exception? Special status codes may require special handling, and it's best not to have the user parse the Exception message.